test/diagonal.jl

```diff
@@ -478,14 +478,14 @@ Random.seed!(1)
         U, s, V = F
         @test map(x -> x.val, Matrix(F)) ≈ map(x -> x.val, Du)
         @test svdvals(Du) == s
-        @test U isa AbstractMatrix{<:Furlong{0}}
-        @test V isa AbstractMatrix{<:Furlong{0}}
+        @test U isa AbstractMatrix{<:AbstractFloat}
```

Comment:
The test failure happens because `D` may be complex as well, and the `<:AbstractFloat` condition doesn't hold. Perhaps we need a branch here on whether `elty <: Real`?

Reply:
Hm, I guess `Union{Real,Complex}` would be sharp enough, since `Furlong <: Number`?
```suggestion
        @test U isa AbstractMatrix{<:Union{Real,Complex}}
```

Another cause of test failures is that when we multiply out the matrices, we go by the most generic matmul routine, which uses `muladd`. That requires a fix, because in unitful multiplication, we don't want the arguments to be promoted, as that may very well fail.